### PR TITLE
web module for serving static files

### PR DESCRIPTION
**Pull request by haliphax**

### CODE REVIEW

File: x84/webmodules/static.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    def GET(self, filename):
+        """ Respond to GET method request. """
+        if not filename:
+            return web.redirect('/www/')
+        myfile = filename
+        myfile = os.path.join(StaticApp.static_root, myfile[1:])
+        myfile = re.sub(r'\.\.|//', '', myfile)
+        if os.path.isfile(myfile):
+            # we're serving a file; use the proper mime type
+            mime = mimetypes.guess_type(myfile)
```
Comment (jquast) [inline x84/webmodules/static.py]: so guess_type fails to guess for any of html, css, or js ??

File: x84/webmodules/static.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            _, ext = os.path.splitext(myfile).lower()
+            if ext == '.html':
+                mime = 'text/html'
+            elif ext == '.css':
+                mime = 'text/css'
+            elif ext == '.js':
+                mime = 'text/javascript'
```
Comment (jquast) [inline x84/webmodules/static.py]:
```python
mime = {'.html': 'text/html''
        '.css': 'text/css',
       '.js': 'text/javasecript'}.get(ext, mime)
```

but only if you really have to, i can't imagine why web.py can't do this

File: x84/webmodules/static.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    static_root = ''
+
+    def GET(self, filename):
+        """ Respond to GET method request. """
+        if not filename:
+            return web.redirect('/www/')
+        myfile = filename
+        myfile = os.path.join(StaticApp.static_root, myfile[1:])
+        myfile = re.sub(r'\.\.|//', '', myfile)
```
Comment (jquast) [inline x84/webmodules/static.py]: can you comment what this regex is for?

File: x84/webmodules/static.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """ Class for serving static files """
+
+    static_root = ''
+
+    def GET(self, filename):
+        """ Respond to GET method request. """
+        if not filename:
+            return web.redirect('/www/')
+        myfile = filename
+        myfile = os.path.join(StaticApp.static_root, myfile[1:])
```
Comment (jquast) [inline x84/webmodules/static.py]: you should not continually re-assign the same variable, "myfile" then becomes "local path of myfile, relative to static root", when then becomes "local path of myfile, relative to static root, with ../ and // stripped entirely"

File: x84/webmodules/static.py (hunk 3 above)
Comment (jquast) [inline x84/webmodules/static.py]: i think these would still fail for windows, you're trusting myfile's http URL "/file/path/to/file" matches our local system's filepath.  You should use ``os.path.join(myfile.split('/'))``

File: x84/webmodules/static.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+""" Static file server web module for x/84 bbs https://github.com/jquast/x84 """
```
Comment (jquast) [inline x84/webmodules/static.py]: please remove "for x/84 bbs https://github.com/jquast/x84" and finish the sentence with a period.

i did a bulk change of this last week, please match

File: x84/webmodules/static.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                mime = 'text/css'
+            elif ext == '.js':
+                mime = 'text/javascript'
+            web.header('Content-Type', mime, unique=True)
+            return open(myfile, 'rb').read()
+        elif os.path.isdir(myfile):
+            # we're serving a directory; try directory/index.html instead
+            if not filename.endswith('/'):
+                return web.redirect('{0}/'.format(filename))
+            return self.GET('{0}index.html'.format(filename))
```
Comment (jquast) [inline x84/webmodules/static.py]: ``os.path.join(myfile, 'index.html')`` will work equally well in both cases

```
>>> os.path.join('a/', 'b')
'a/b'
>>> os.path.join('a', 'b')
'a/b'
```

File: x84/webmodules/static.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    """ Expose our REST API. Run only once on server startup. """
+    from x84.bbs.ini import get_ini
+
+    # determine document root for web server
+    static_root = (get_ini('web', 'document_root')
+                   or os.path.join(get_ini('system', 'scriptpath'), 'www'))
+    StaticApp.static_root = static_root
+
+    return {
+        'urls': ('/www(/.*)?', 'static'),
```
Comment (jquast) [inline x84/webmodules/static.py]: does this mean static files are only relative to a **/www/**static-path.js ?

if we have to have it, then maybe it should be 'static', both as the /static folder, and scriptpath/webmodules/static module folder

**Comment by haliphax** [comment]
Well, I think I'm going to see if we can use `cherrypy` directly without web.py being involved. It hooks URL patterns to classes like `web.py`. Context is a tiny bit harder to work with, but we would be able to have complete control over where its tested method for serving static files is serving them from.

File: x84/webmodules/static.py (hunk 1 above)
Comment (haliphax) [inline x84/webmodules/static.py]: Correct. It sends them as text/plain.

File: x84/webmodules/static.py (hunk 6 above)
Comment (haliphax) [inline x84/webmodules/static.py]: It's not just about serving the right file -- it's also about telling the browser to go to the right URL. If we ensure that directories always end with a forward slash, you won't get duplicate entries if you want to use analytics or anything on the log. Additionally, it's not the OS's path separator we want to use -- it's a URL path separator, which is always '/', regardless of the operating system.

File: x84/webmodules/static.py (hunk 7 above)
Comment (haliphax) [inline x84/webmodules/static.py]: Will handle this in my re-push.

**Pull request merged**